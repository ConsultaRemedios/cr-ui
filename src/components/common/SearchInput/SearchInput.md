SearchInput is a search input component.
This component just search on enter or on click in search button.

```js
const showClearButton = () => {
  return Boolean(window.location.search);
} ;

const onClickClearButton = () => {
  window.location.assign(`${window.location.pathname}`);
};

const getValue = () => {
  let searchValue = window.location.search.substring(7);
  let value = searchValue.replace(/\+/g, ' ');
  return value;
}

const onSubmitInput = ({ value }) => {
  const searchValue = value.replace(/ /g, '+');
  window.location.assign(`${window.location.pathname}?termo=${searchValue}`);
};

<div>
  <SearchInput
    placeholder="Buscando um pedido especÃ­fico?"
    :value="getValue()"
    :show-clear-button="showClearButton()"
    @click-clear-button=onClickClearButton
    @submit="onSubmitInput"
  />
</div>
```

Another example

```js
new Vue({
  data() {
    return {
      ui: {
        showClearButton: false,
      }
    };
  },
  methods: {
    toggleClearButton() {
      this.ui.showClearButton = !this.ui.showClearButton;
    },

    onClickClearButton() {
      alert('Clicked clear button');
    },

    onSubmit({ value }) {
      alert(`Submitted with value: ${value}`);
    },
  },
  template: `
    <div>
      <BaseButton
        :style="{ marginBottom: '10px' }"
        @click="toggleClearButton"
      >
        Toggle Clear Button
      </BaseButton>

      <SearchInput
        placeholder="Digite algo"
        :show-clear-button="ui.showClearButton"
        @click-clear-button=onClickClearButton
        @submit="onSubmit"
      />
    </div>
  `
})
```
