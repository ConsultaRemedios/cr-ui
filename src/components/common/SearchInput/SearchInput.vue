<template>
  <div :class="$style.container">
    <div
      v-if="showClearButton"
      :class="$style.iconWrapper"
      @click="onClickClearButton"
    >
      <BaseIcon
        :id="cancelIcon.id"
        :class="$style.cancelIcon"
      />
    </div>
    <form
      :class="$style.form"
      @submit.prevent="onSearch"
    >
      <input
        type="search"
        :class="$style.input"
        :value="inputValue"
        :placeholder="placeholder"
        @input="onInput"
      >
      <button :class="$style.buttonSearch">
        <BaseIcon
          :id="searchIcon.id"
          :class="$style.searchIcon"
        />
      </button>
    </form>
  </div>
</template>

<script>
  import BaseIcon from '../BaseIcon';
  import search from '../../../icons/search.icon.svg';
  import cancel from '../../../icons/cancel.icon.svg';

  export default {
    name: 'SearchInput',

    props: {
      value: {
        type: String,
        default: '',
      },

      placeholder: {
        type: String,
        default: 'Buscar...',
      },

      showClearButton: {
        type: Boolean,
        default: false,
      },
    },

    data() {
      return {
        inputValue: '',
      };
    },

    methods: {
      onInput({ target }) {
        this.inputValue = target.value;
      },

      onSearch() {
        this.$emit('submit', { value: this.inputValue });
      },

      onClickClearButton() {
        this.$emit('click-clear-button');
      },
    },

    computed: {
      searchIcon() {
        return search;
      },

      cancelIcon() {
        return cancel;
      },
    },

    created() {
      if (this.value) {
        this.inputValue = this.value;
      }
    },

    components: {
      BaseIcon,
    },
  };
</script>

<style module>
  .container {
    position: relative;
  }

  .form {
    border: 1px solid #DADADA;
    border-radius: 3px;
  }

  .input {
    width: calc(100% - 20px);
    height: 45px;
    padding: 11px 15px;
    border: none;
    color: #333333;
    font-size: 15px;
    background-color: transparent;
    -webkit-appearance: none;
  }

  .input[placeholder] {
    text-overflow: ellipsis;
  }

  .input:focus {
    border-width: 2px;
    border-color: #00AAE5;
  }

  .iconWrapper + form > .input {
    padding-left: 40px;
  }

  .input::placeholder {
    color: #AAAAAA;
    opacity: 1;
  }

  .input::-webkit-search-cancel-button {
    display: none;
  }

  .input::-ms-clear{
    display: none;
  }

  .cancelIcon {
    position: absolute;
    top: 50%;
    left: 10px;
    transform: translateY(-50%);
    width: 20px;
    height: 20px;
    color: #D30B3F;
    cursor: pointer;
  }

  .buttonSearch {
    position: absolute;
    top: 50%;
    right: 5px;
    transform: translateY(-50%);
    background-color: transparent;
    border: none;
  }

  .searchIcon {
    width: 20px;
    height: 20px;
    color: #999999;
    cursor: pointer;
  }
</style>
