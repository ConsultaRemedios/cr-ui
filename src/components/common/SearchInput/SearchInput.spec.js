import { mount } from '@vue/test-utils';
import SearchInput from './SearchInput.vue';

describe('SearchInput component', () => {
  const mountSearchInput = (props = {}, methods) => {
    const propsData = { ...defaultProps, ...props };
    return mount(SearchInput, { propsData, methods });
  };

  const defaultProps = {
    showClearButton: true,
    placeholder: 'Buscar...',
    value: 'Some text to search',
  };

  describe('Props', () => {
    describe('#showClearButton', () => {
      it('returns an empty array when the prop is not passed to it', () => {
        const wrapper = mount(SearchInput);

        expect(wrapper.vm.showClearButton).toEqual(false);
      });
    });

    describe('#value', () => {
      it('returns an empty string when the prop is not passed to it', () => {
        const wrapper = mount(SearchInput);

        expect(wrapper.vm.value).toEqual('');
      });
    });

    describe('#placeholder', () => {
      it('returns the default placeholder when the prop is not passed to it', () => {
        const wrapper = mount(SearchInput);

        expect(wrapper.vm.placeholder).toEqual('Buscar...');
      });
    });
  });

  describe('matches the snapshot', () => {
    let wrapper;

    it('when "showClearButton" is false', () => {
      wrapper = mountSearchInput({ showClearButton: false });

      expect(wrapper.element).toMatchSnapshot();
    });

    it('when "showClearButton" is true', () => {
      wrapper = mountSearchInput({ showClearButton: true });

      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('Event listeners', () => {
    describe('when the input value changes', () => {
      let wrapper;
      let onInput;

      beforeEach(() => {
        onInput = jest.fn();
        wrapper = mountSearchInput({ value: 'another text' }, { onInput });

        expect(onInput).not.toHaveBeenCalled();

        wrapper.find('input').trigger('input');
      });

      it('calls #onInput', () => {
        expect(onInput).toHaveBeenCalledTimes(1);
      });

      it('calls #onInput passing the proper param', () => {
        expect(onInput.mock.calls[0][0].target.value).toEqual('another text');
      });
    });
  });

  describe('When passing "value" prop', () => {
    let wrapper;

    beforeAll(() => {
      wrapper = mountSearchInput({ value: 'another text' });
    });

    it('sets "inputValue" data with the value prop', () => {
      expect(wrapper.vm.inputValue).toEqual('another text');
    });
  });

  describe('Methods', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = mountSearchInput({ value: 'changed input text' });
    });

    describe('#onInput', () => {
      describe('When is called', () => {
        beforeEach(() => {
          const event = {
            target: { value: 'changed input text' }
          };

          wrapper.vm.onInput(event);
        });

        it('changes the "inputValue" according to "target.value"', () => {
          expect(wrapper.vm.inputValue).toEqual('changed input text');
        });
      });
    });

    describe('#onSearch', () => {
      describe('When is called', () => {
        it('emits "submit" event', () => {
          jest.spyOn(wrapper.vm, '$emit');

          wrapper.find('button').trigger('submit');
          let emitted = wrapper.vm.$emit.mock.calls[0];

          expect(emitted[0]).toEqual('submit');
        });

        it('the emitted was called with the proper value', () => {
          jest.spyOn(wrapper.vm, '$emit');

          wrapper.find('button').trigger('submit');
          let emitted = wrapper.vm.$emit.mock.calls[0];

          expect(emitted[1]).toMatchObject({ value: 'changed input text' });
        });
      });
    });

    describe('#onClickClearButton', () => {
      describe('When is called', () => {
        it('emits "click-clear-button" event', () => {
          jest.spyOn(wrapper.vm, '$emit');

          wrapper.find('.iconWrapper').trigger('click');
          const emitted = wrapper.vm.$emit.mock.calls[0];

          expect(emitted[0]).toEqual('click-clear-button');
        });
      });
    });
  });

  describe('Computed', () => {
    describe('#searchIcon', () => {
      it('returns search icon', () => {
        const wrapper = mountSearchInput();

        expect(wrapper.vm.searchIcon).toEqual({
          id: 'search.icon.svg',
          viewbox: '0 0 24 24',
        });
      });
    });

    describe('#cancelIcon', () => {
      it('returns cancel icon', () => {
        const wrapper = mountSearchInput();

        expect(wrapper.vm.cancelIcon).toEqual({
          id: 'cancel.icon.svg',
          viewbox: '0 0 24 24',
        });
      });
    });
  });
});
