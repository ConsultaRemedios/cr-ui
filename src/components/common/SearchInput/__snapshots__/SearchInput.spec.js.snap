// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput component matches the snapshot when "showClearButton" is false 1`] = `
<div
  class="container"
>
  <!---->
   
  <form
    class="form"
  >
    <input
      class="input"
      placeholder="Buscar..."
      type="search"
    />
     
    <button
      class="buttonSearch"
    >
      <svg
        class="icon searchIcon"
        viewBox="0 0 24 24"
      >
        <use
          xlink:href="#search.icon.svg"
        />
      </svg>
    </button>
  </form>
</div>
`;

exports[`SearchInput component matches the snapshot when "showClearButton" is true 1`] = `
<div
  class="container"
>
  <div
    class="iconWrapper"
  >
    <svg
      class="icon cancelIcon"
      viewBox="0 0 24 24"
    >
      <use
        xlink:href="#cancel.icon.svg"
      />
    </svg>
  </div>
   
  <form
    class="form"
  >
    <input
      class="input"
      placeholder="Buscar..."
      type="search"
    />
     
    <button
      class="buttonSearch"
    >
      <svg
        class="icon searchIcon"
        viewBox="0 0 24 24"
      >
        <use
          xlink:href="#search.icon.svg"
        />
      </svg>
    </button>
  </form>
</div>
`;
